chrome/content/zotero/elements/bubbleInput.js

```diff
@@ -0,0 +1,667 @@
+						break;
+					}
+				}
+			}
+			if (lastBubble !== null) {
+				lastBubble = bubbles[lastBubble];
+			}
+			return lastBubble;
+		},
+
+		notifyDialog(eventType, data = {}) {
+			let event = new CustomEvent(eventType, {
+				bubbles: true,
+				detail: data
+			});
+			this.bubbleInput.dispatchEvent(event);
+		},
+		// Determine if keypress event is on a printable character.
+		/* eslint-disable array-element-newline */
+		isKeypressPrintable(event) {
+			if (event.ctrlKey || event.metaKey || event.altKey) return false;
+			// If it's a single character, for latin locales it has to be printable
+			if (event.key.length === 1) {
+				return true;
+			}
+			// Otherwise, check against a list of common control keys
+			let nonPrintableKeys = [
+				'Enter', 'Escape', 'Backspace', 'Tab',
+				'ArrowLeft', 'ArrowRight', 'ArrowUp', 'ArrowDown',
+				'Home', 'End', 'PageUp', 'PageDown',
+				'Delete', 'Insert',
+				'F1', 'F2', 'F3', 'F4', 'F5', 'F6', 'F7', 'F8', 'F9', 'F10', 'F11', 'F12',
+				'Control', 'Meta', 'Alt', 'Shift', 'CapsLock'
+			];
+			/* eslint-enable array-element-newline */
+		
+			return !nonPrintableKeys.includes(event.key);
+		},
+
+		buildBubbleString({ citationItem, zoteroItem }) {
+			// Creator
+			var title;
+			var str = zoteroItem.getField("firstCreator");
+			
+			// Title, if no creator (getDisplayTitle in order to get case, e-mail, statute which don't have a title field)
+			title = zoteroItem.getDisplayTitle();
+			title = title.substr(0, 32) + (title.length > 32 ? "…" : "");
+			if (!str && title) {
+				str = Zotero.getString("punctuation.openingQMark") + title + Zotero.getString("punctuation.closingQMark");
+			}
+			else if (!str) {
+				str = Zotero.getString("integration-citationDialog-bubble-empty");
+			}
+			
+			// Date
+			var date = zoteroItem.getField("date", true, true);
+			if (date && (date = date.substr(0, 4)) !== "0000") {
+				str += ", " + parseInt(date);
+			}
+			
+			// Locator
+			if (citationItem.locator) {
+				// Try to fetch the short form of the locator label. E.g. "p." for "page"
+				// If there is no locator label, default to "page" for now
+				let label = (Zotero.Cite.getLocatorString(citationItem.label || 'page', 'short') || '').toLocaleLowerCase();
+				
+				str += `, ${label} ${citationItem.locator}`;
+			}
+			
+			// Prefix
+			if (citationItem.prefix && Zotero.CiteProc.CSL.ENDSWITH_ROMANESQUE_REGEXP) {
+				let prefix = citationItem.prefix.substr(0, 10) + (citationItem.prefix.length > 10 ? "…" : "");
+				str = prefix
+					+ (Zotero.CiteProc.CSL.ENDSWITH_ROMANESQUE_REGEXP.test(citationItem.prefix) ? " " : "")
+					+ str;
+			}
+			
+			// Suffix
+			if (citationItem.suffix && Zotero.CiteProc.CSL.STARTSWITH_ROMANESQUE_REGEXP) {
+				let suffix = citationItem.suffix.substr(0, 10) + (citationItem.suffix.length > 10 ? "…" : "");
+				str += (Zotero.CiteProc.CSL.STARTSWITH_ROMANESQUE_REGEXP.test(citationItem.suffix) ? " " : "") + suffix;
+			}
+			
+			return str;
+		},
```

Comment:
Can we do this in citationDialog.js?

Reply:
Yeah, for sure! I moved it to `helpers.js` to keep `citationDialog.js` relatively clean. Then, we're using that function to make a new string before refreshing bubble-input and passing it together with other parameters when `.refresh` from `bubble-input` is called.